### ISSUE TITLE
Give option to load custom stuff from the checkpoint instance manually.

### ISSUE THREAD

**Issue by FeryET** [PyTorchLightning/pytorch-lightning#10188]
### Discussed in https://github.com/PyTorchLightning/pytorch-lightning/discussions/10172

<div type='discussions-op-text'>

<sup>Originally posted by **FeryET** October 27, 2021</sup>
I want to create an inference model, and I have multiple datasets. For this issue, I want to save some settings about my DataModule and indicate what is the test indices, what is the dataset location and etc.

How can I add these to model_checkpoint?</div>

------------------------------------------------

Currently I can add custom fields to my checkpoint using `on_save_checkpoint` callback that both `LightningDataModule` and `LightningModule` override. But the problem is that while I can load my model's weights (and probably states) using `LightningModule.load_from_checkpoint` I have no such options to load the custom things I added on `LightningDataModule.on_save_checkpoint` back to the `LightningDataModule` instance I have newly created. For example the indices of train, test, and validation sets on the whole data is a field I would like to keep in my checkpoint, and I want to load them later in my datamodule to have reproducable splitted sets. Or even the configuration or random seed or whatever. 

Currently there is a dirty hack that I have to include my datamodule in the main model (`LightningModule`) for it to be saved and loaded. But this is bloated and has major conflicts with the separation of concerns.

cc @borda

**Comment by FeryET** [comment]
> I'd be very interested in an elegant solution to this as well. Currently, there are many moving pieces that remain crucial to reproducible workflows but are difficult or inconvenient to always attach to the LightningModule.
> 
> Could you subclass the basic Lightning Trainer and override the load_from_checkpoint method to wrap internal Module and Datamodule-specific logic?
> 
> e.g.
> 
> ```
> class Trainer(pl.Trainer):
> 
> 	@classmethod
> 	def load_from_checkpoint(cls, ckpt_path):
> 		ckpt = torch.load(ckpt_path)
> 		datamodule = pl.DataModule.load_from_checkpoint(ckpt["datamodule"])
> 		# ... insert datamodule-specific logic here
> 		pl_module = pl.LightningModule.load_from_checkpoint(ckpt["pl_module"])
> 		return datamodule, pl_module
> ```

I think giving the option to people to load Model and DataModule, or whatever they have added to the checkpoint seperately, or loading them all via calling the trainer is the best option. Currently I use the `load` function from `cloud_io.py` file, which locally uses a simple `torch.load` backend, but there are 2 things:

1. I don't know if this scales to multi-gpu cloud solutions when each cpu has its own checkpoint.
2. This is too much verbosity and I had to dig really dip to find this function.
